Bullet-dodging game/Bullet-Dodging Game in Processing.md

```diff
@@ -0,0 +1,2191 @@
+
+```
+void keyPressed() {
+  if (key == 32) {  // keyCode for spacebar
+    if (state == 0) {  // start the game if it hasn't started
+      state = 1;
+    } else if (state == 2) {  // after game over, reinitialize the variables if the game is to be restarted
+      state = 1;
+      hp = 6;
+      bullet_num = 60;
+      score = 0;
+      enemies_destroyed = 0;
+      hit = 0;
+      total = 0;
+      time = 0;
+      last_time = 0;
+      player = new Player();
+      life = new Life_Icon();
+      bul = new Bullet_Icon();
+      my_bullets = new ArrayList<Bullet>();
+      bullets = new ArrayList<Bullet>();
+      enemies = new ArrayList<Enemy>();
+    } else {  //pause or resume the game during the game
+      pause = !pause;
+    }
+  }
+}
+```
+
+In addition, we don't want the player to take advantage of shooting in the paused state, so in the `mousePressed()` function we'll only call `player.shoot()` when the game is on:
+
+```
+void mousePressed() {
+  if (state == 1 && ! pause) {
+    player.shoot();
+  }
+}
+```
+
+Now we can use the two state variables to control the flow of the game. Below is the structure of the code in `draw()` that gives us the ability to pause and restart the game.
+
+First, we check if `state == 0` (the game has not started yet). If so, we only need to call `show()` without `update()`.
+
+```
+if (state == 0) {  // before the game starts
+  background(background);
+  image(obstacle, 500, 700, 200, 200);
+  player.show();
+  life.show();
+  bul.show();
+  show_status();
+  fill(0);
+  textSize(60);
+  text("Press SPACE to start/pause", 200, 400);
+}
+```
+
+Then we check if `state == 1` (the game has started but not finished). We put our previous code in `draw()` in the following block that will run when the game is not paused. On the other hand, we simply show a line of prompt without performing any other functions when the game is paused.
+
+```
+else if (state == 1) {  // during the game
+  if (!pause) {  // perform all the functions when the game is not paused
+  	// previous code in draw()
+  } else {  // if the game is paused, stop updating any data
+    fill(0);
+    textSize(60);
+    text("Press SPACE to resume", 270, 400);
+  }
+}    
+```
+
+Also, if the game is over, we can choose to print a message to the screen. We can also record the highest score by declaring a global variable `high_score == 0` initially.
+
+```
+// update the highest score
+if (score > high_score) {
+  high_score = score;
+}
+// dark background at the center of the screen
+noStroke();
+fill(0, 80);
+rect(220, 270, 800, 400, 50);
+// show game over and the highest score
+fill(0);
+textSize(60);
+text("Game Over", 450, 380);
+textSize(40);
+text("High score: " + high_score, 445, 470);
+textSize(60);
+text("Press SPACE to restart", 300, 580);
+```
+
+After viewing the high score, we can now restart the game after game over!
+
+![](images/image3.png)
+
+
+

```

Comment:
since you've put in code dumps already, I would add a final one to the end here for people to compare against